### PR TITLE
Update R2/KJ to avoid mod-after-put when using ArrayBuffer

### PR DESCRIPTION
**Pull request by jasnell**
Prior to this change, R2/KV would allow post-call modification of the ArrayBuffer. This updates it to copy by default and detach with enablement of a compat flag.

### CODE REVIEW

**Comment by Frederik-Baetens** [comment]
Why is the compat flag needed? Was it not possible before that the data would be sent over the wire before any post-call modification could take place? If so, why would it not be acceptable to now always have that happen instead of just sometimes?

**Comment by vlovich** [comment]
@jasnell can we instrument this with Sentry and see if anything is actually triggering this codepath? If we run for a while and no one is using it, it should be pretty safe to just adjust the default to not need a compat flag right?

**Review by vlovich — APPROVED** [review]
LGTM for KV. Skimmed R2 looks OK but someone from that team should approve.

**Comment by jasnell** [comment]
So there are actually three modes here.

The old mode... which this PR removes, is that post-call changes are picked up in the put. e.g.

```js
const u8 = new Uint8Array(10);
u8.fill(1);
const p = bucket.put(..., u8);
u8.fill(2);  // the bucket sees [2,2,2,....] when it should see [1,1,1,....]
await p;
```

The new default mode is that post-call changes are ignored because the ArrayBuffer is copied. However, the passed in ArrayBuffer/Uint8Array *is still usable* by user code.

```js
const u8 = new Uint8Array(10);
u8.fill(1);
const p = bucket.put(..., u8);
u8.fill(2);  // the bucket sees [1,1,1,1....] but the u8 is still directly usable
console.log(u8.byteLength);  // still prints 10
await p;
```
This comes at the cost of the potentially expensive copy.

With the compatibility flag, we avoid the copy but post-call, the Uint8Array is no longer usable, which could break user code.

```js
// with the compat flag set...
const u8 = new Uint8Array(10);
u8.fill(1);
const p = bucket.put(..., u8);
u8.fill(2);  // the bucket sees [1,1,1,1....] but the u8 is NO LONGER directly usable
console.log(u8.byteLength);  // prints 0!!
await p;
```

The best we could do with the sentry log is detect how often someone passes in an ArrayBuffer. We cannot detect if they are continuing to use the ArrayBuffer after the put. This could break, for instance, if someone is doing additional logging of the buffer after the put, or directing it somewhere else.

```js
// with the compat flag set...
const u8 = new Uint8Array(10);
u8.fill(1);
const p = bucket.put(..., u8);
u8.fill(2);  // does nothing, the bucket sees [1,1,1,1....] but the u8 is NO LONGER directly usable
console.log(u8);  // prints nothing!!!
await p;
```

We need the compat flag because of this side effect

**Review by Frederik-Baetens — APPROVED** [review]
+1 for R2

File: src/workerd/api/r2-rpc.c++ [hunk 1]
```diff
@@ -162,9 +163,17 @@ kj::Promise<R2Result> doR2HTTPPutRequest(jsg::Lock& js, kj::Own<kj::HttpClient>
...
         KJ_REQUIRE(streamSize == nullptr);
       }
-      KJ_CASE_ONEOF(data, kj::Array<kj::byte>) {
+      KJ_CASE_ONEOF(data, jsg::BufferSource) {
         expectedBodySize = data.size();
         KJ_REQUIRE(streamSize == nullptr);
+        // We want to either detach or copy the given ArrayBuffer/TypedArray
+        // in order to prevent the possibility of post-call modification.
+        if (featureFlags.getDetachArrayBufferOnPut()) {
+          maybeBackingStore = data.detach(js);
```
Comment (Frederik-Baetens) [inline src/workerd/api/r2-rpc.c++:172]: Is it possible for the JS code to modify this buffer before this code is reached? Or is it impossible for the c++ code to be pre-empted between the function call & this code?

File: src/workerd/io/compatibility-date.capnp [hunk 2]
```diff
@@ -254,4 +254,13 @@ struct CompatibilityFlags @0x8f8c1b68151b6cef {
   # WHATWG URL Standard, leading to a number of issues reported by users. Unfortunately,
   # the specCompliantUrl flag did not contemplate the redirect usage. This flag is
   # specifically about the usage in a redirect().
+
+  detachArrayBufferOnPut @24 :Bool
+      $compatEnableDate("2023-03-01")
+      $compatEnableFlag("detach_arraybuffer_on_kv_r2_put")
+      $compatEnableFlag("copy_arraybuffer_on_kv_r2_put");
```
Comment (harrishancock) [inline src/workerd/io/compatibility-date.capnp:261]:
```suggestion
      $compatDisableFlag("copy_arraybuffer_on_kv_r2_put");
```

File: src/workerd/api/r2-rpc.c++ (hunk 1 above)
Comment (harrishancock) [reply src/workerd/api/r2-rpc.c++:172]: As long as this function is called synchronously, JS can't modify the buffer before this code is reached.

File: src/workerd/jsg/buffersource.c++ [hunk 3]
```diff
@@ -117,6 +128,14 @@ BackingStore BufferSource::detach(Lock& js) {
   return kj::mv(backingStore);
 }
 
+BackingStore BufferSource::cloneBackingStore(Lock& js) {
+  auto backingStore =
+      kj::mv(JSG_REQUIRE_NONNULL(maybeBackingStore,
```
Comment (harrishancock) [inline src/workerd/jsg/buffersource.c++:133]: Is the `kj::mv()` intended? It looks like this function will leave the BufferSource in an unusable state, but perhaps that's intended.

Either way, a good test case to add to buffersource-test.c++. ;)

Comment (jasnell) [reply src/workerd/jsg/buffersource.c++:133]: No, good catch. Cut'n'paste issue.

**Pull request merged**